service/internal/clientpair/clientpair.go

```diff
@@ -0,0 +1,95 @@
+package clientpair
+
+import (
+	"context"
+	"strings"
+
+	"github.com/giantswarm/apiextensions-application/api/v1alpha1"
+	"github.com/giantswarm/helmclient/v4/pkg/helmclient"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+
+	"github.com/giantswarm/chart-operator/v2/service/controller/chart/key"
+)
+
+const (
+	// privateNamespace defines GS-protected namespace the prvHelmClient
+	// is meant for. For App CRs created outside this namespace, the
+	// pubHelmClient should be used
+	privateNamespace = "giantswarm"
+
+	// WC App Operators are the only cases where the elevated prvHelmClient
+	// has to be used outside the `giantswarm` namespace. It is due to these
+	// apps installing reources the `default:automation` does not have access to.
+	// However, existance of the App CR for unique App Operator outside the
+	// `giantswarm` namespace opens a door to scenarios described here:
+	// https://github.com/giantswarm/giantswarm/issues/22100.
+	// The appOperatorChart defines legal prefix for what elevated Helm Client
+	// can install outside the `giantswarm` namespace
+	appOperatorChart = "https://giantswarm.github.io/control-plane-catalog/app-operator"
+)
+
+type ClientPairConfig struct {
+	Logger micrologger.Logger
+
+	PrvHelmClient helmclient.Interface
+	PubHelmClient helmclient.Interface
+}
+
+type ClientPair struct {
+	logger micrologger.Logger
+
+	prvHelmClient helmclient.Interface
+	pubHelmClient helmclient.Interface
+}
+
+func NewClientPair(config ClientPairConfig) (*ClientPair, error) {
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+	if config.PrvHelmClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.PrvHelmClient must not be empty", config)
+	}
+
+	cp := &ClientPair{
+		logger: config.Logger,
+
+		prvHelmClient: config.PrvHelmClient,
+		pubHelmClient: config.PubHelmClient,
+	}
+
+	return cp, nil
+}
+
+// Get determines which client to use based on the namespace the corresponding App CR
+// is located in. For Workload Cluster, chart operator is permitted to operate under
+// cluster-wide permissions, so there is only prvHelmClient used
```

Comment:
```suggestion
// cluster-wide permissions, so there is only prvHelmClient used.
```

Turbo nitpick but sentences in godoc should always end with a `.` and they basically never do in this PR :P